src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.State.cs

```diff
@@ -90,6 +106,25 @@ protected State(ValueSource<Optional<Compilation>>? compilation, Compilation? de
                         ? new WeakValueSource<Compilation>(compilation)
                         : (ValueSource<Optional<Compilation>>)new ConstantValueSource<Optional<Compilation>>(compilation);
                 }
+
+                public static ConditionalWeakTable<ISymbol, ISymbol> GetCompilationAssembliesAndModules(Compilation compilation)
+                {
+                    var result = new ConditionalWeakTable<ISymbol, ISymbol>();
+
+                    var compAssembly = compilation.Assembly;
+                    result.Add(compAssembly, compAssembly);
```

Comment:
>compAssembly, compAssembly [](start = 31, length = 26)

I don't understand why we are mapping the symbols onto themselves.

Reply:
there's no ConditionalWeakSet<T>.  So this just simulates that.  i just needed some value to put there.  I can comment appropriately.

Effectively we just want to ensure that if *someone* is holding these assembly symbols around, that we can then use that symbol to map back to an appropriate project where it could have come from.  

So we want it to be:
1. conditional.  so we're not rooting these objects.
2. all the potential assembly/module symbols for a compilation.

`CWT<ISymbo, ...>` fits the bill.  for expedince i just used the key as the value.

Reply:
Are you using it to ask a yes/no question "is this object alive"?

Reply:
I see that's the case:

```C#
                        foreach (var (id, state) in _solutionState.ProjectStates)
                        {
                            var tracker = _solutionState.GetCompilationTracker(id);
                            if (tracker.ContainsAssemblyOrModule(symbol))
                            {
                                projectId = id;
                                break;
                            }
                        }
```

Couldn't we instead use `CWT<ISymbol, ProjectId>` on SolutionState and avoid enumerating the project states?

Reply:
> there's no ConditionalWeakSet. 

I too had the "what is this map?" comment of @tmat. If you checked in a "WeakSet" that just wrapped the CWT so implementation isn't leaking to the higher concerns I wouldn't complain.

Reply:
> Are you using it to ask a yes/no question "is this object alive"?

Not exactly.  The symbol must be alive if someone is asking about it.  It's more like:

1. did this symbol come from this project?
2. how do answer that without rooting the symbol?  i.e. if no one is holdin gonto it, we want to let the GC clean it up.

Reply:
> Couldn't we instead use CWT<ISymbol, ProjectId> on SolutionState and avoid enumerating the project states?

I tried that, but it's much more difficult because any project change cause us to create a new SolutionState.   We would be unable to share the previous SolutionState cache with the next (unlike with ProjectState, where we only need ot recompute for that project).

I'm really not certain how that would work (and seems like it would actually add a *lot* of complexity).

Reply:
> I too had the "what is this map?" comment of @tmat. If you checked in a "WeakSet" that just wrapped the CWT so implementation isn't leaking to the higher concerns I wouldn't complain.

That sounds very reasonable.

Reply:
> We would be unable to share the previous SolutionState cache with the next 

Actually, why not make the CWT static? The GC will clean it up as old symbols are collected.

If you still prefer the CWT on ProjectState I think it'd be better having `CWT<ISymbol, object>` where object is a singleton empty object instance because the value is not relevant.

Reply:
> (as we update the solution to the next we reuse the whole tabl

That could lead to invalid tables.  For example, it we share one table, and it says IAssembly Foo came from Project A, that coudl be invalid if IAssumbly Foo is removed from Proejct A and added to Project B.  

> If you still prefer the CWT on ProjectState I think it'd be better having CWT<ISymbol, object> where object is a singleton empty object instance because the value is not relevant.

Yup.  I made that change last night :)  We actually just map from IAssemblYSymbol to null now.

Reply:
I see. Makes sense then.